src/core/Akka.Persistence/Eventsourced.Recovery.cs

```diff
@@ -65,52 +83,64 @@ private EventsourcedState RecoveryStarted(long maxReplays)
 
             return new EventsourcedState("recovery started - replay max: " + maxReplays, true, (receive, message) =>
             {
-                if (message is LoadSnapshotResult)
+                try
                 {
-                    var res = (LoadSnapshotResult)message;
-                    timeoutCancelable.Cancel();
-                    if (res.Snapshot != null)
+                    if (message is LoadSnapshotResult res)
                     {
-                        var snapshot = res.Snapshot;
-                        LastSequenceNr = snapshot.Metadata.SequenceNr;
-                        // Since we are recovering we can ignore the receive behavior from the stack
-                        base.AroundReceive(recoveryBehavior, new SnapshotOffer(snapshot.Metadata, snapshot.Snapshot));
-                    }
+                        timeoutCancelable.Cancel();
+                        if (res.Snapshot != null)
+                        {
+                            var snapshot = res.Snapshot;
+                            LastSequenceNr = snapshot.Metadata.SequenceNr;
+                            // Since we are recovering we can ignore the receive behavior from the stack
+                            base.AroundReceive(recoveryBehavior, new SnapshotOffer(snapshot.Metadata, snapshot.Snapshot));
+                        }
 
-                    ChangeState(Recovering(recoveryBehavior, timeout));
-                    Journal.Tell(new ReplayMessages(LastSequenceNr + 1L, res.ToSequenceNr, maxReplays, PersistenceId, Self));
-                }
-                else if (message is LoadSnapshotFailed)
-                {
-                    var res = (LoadSnapshotFailed)message;
-                    timeoutCancelable.Cancel();
-                    try
+                        ChangeState(Recovering(recoveryBehavior, timeout));
+                        Journal.Tell(new ReplayMessages(LastSequenceNr + 1L, res.ToSequenceNr, maxReplays, PersistenceId, Self));
+                    }
+                    else if (message is LoadSnapshotFailed failed)
                     {
-                        OnRecoveryFailure(res.Cause);
+                        timeoutCancelable.Cancel();
+                        try
+                        {
+                            OnRecoveryFailure(failed.Cause);
+                        }
+                        finally
+                        {
+                            Context.Stop(Self);
+                        }
                     }
-                    finally
+                    else if (message is RecoveryTick)
```

Comment:
Scala version has `case RecoveryTick(true)`, so we also should check `if (message is RecoveryTick tick && tick.Snapshot == true)`